### PR TITLE
[Plutus Core] [Plutus IR] [Uniques] Do not keep closed things in 'Quote'

### PR DESCRIPTION
**Pull request by effectfully**
This removes a lot of `Quote`s and `rename`s. I'll leave some comments below.

I'm planning to write on out uniques handling policies in a future PR.

### CODE REVIEW

File: language-plutus-core/examples/Language/PlutusCore/Examples/Data/InterList.hs [hunk 1]
```diff
@@ -52,9 +51,9 @@ getBuiltinInterList = do
         . TyFun () (mkIterTyFun () [TyVar () a, TyVar () b, interlistBA] $ TyVar () r)
         $ TyVar () r
 
-getBuiltinInterNil :: Quote (Term TyName Name ())
-getBuiltinInterNil = rename =<< do
-    RecursiveType interlist wrapInterList <- getBuiltinInterList
+interNil :: Term TyName Name ()
+interNil = runQuote $ do
```
Comment (effectfully) [inline language-plutus-core/examples/Language/PlutusCore/Examples/Data/InterList.hs]: It is safe to call `runQuote` over a type or a term that must be closed (i.e. when the type/term not being closed implies there is a bug).

File: language-plutus-core/examples/Language/PlutusCore/Examples/Data/TreeForest.hs [hunk 2]
```diff
@@ -74,6 +74,10 @@ using this representation:
     Forest = AsForest TreeForest
 -}
 
+{- Note [Renaming]
+We do renaming in this module, because we normalize things and this requires renaming.
+-}
```
Comment (effectfully) [inline language-plutus-core/examples/Language/PlutusCore/Examples/Data/TreeForest.hs]: So we do invoke type normalization not only from the type checker. Hence the normalizer should be split into two modules (internals, user-facing API) like the renamer and the type checker.

File: language-plutus-core/generators/Language/PlutusCore/Generators/Interesting.hs [hunk 3]
```diff
@@ -6,39 +6,40 @@
...
 
-import           Control.Monad.Morph
 import           Data.List                                (genericIndex)
 import           Hedgehog                                 hiding (Size, Var)
 import qualified Hedgehog.Gen                             as Gen
 import qualified Hedgehog.Range                           as Range
 
 -- | The type of terms-and-their-values generators.
-type TermGen size a = GenT Quote (TermOf (TypedBuiltinValue size a))
+type TermGen size a = Gen (TermOf (TypedBuiltinValue size a))
```
Comment (effectfully) [inline language-plutus-core/generators/Language/PlutusCore/Generators/Interesting.hs]: Whenever we cross language boundaries, i.e.

1. generate a PLC term that has a Haskell counterpart
2. make a PLC term out of a Haskell value
3. make a Haskell value out of a PLC term

we can safely ignore `Quote`, because everything must be closed.

File: language-plutus-core/generators/Language/PlutusCore/Generators/Internal/Entity.hs [hunk 4]
```diff
@@ -242,12 +248,12 @@ genTerm genBase = go where
...
 -- There are still like a half of terms that fail with out-of-bounds errors being evaluated.
-genTermLoose :: TypedBuiltinGenT Quote
+genTermLoose :: Monad m => TypedBuiltinGenT m
 genTermLoose = genTerm genTypedBuiltinSmall typedBuiltinNames 4
 
 -- | Generate a 'TypedBuiltin' and a 'TermOf' of the corresponding type,
 -- attach the 'TypedBuiltin' to the value part of the 'TermOf' and pass that to a continuation.
 withAnyTermLoose
-    :: (forall a. TermOf (TypedBuiltinValue Size a) -> GenT Quote c) -> GenT Quote c
+    :: Monad m => (forall a. TermOf (TypedBuiltinValue Size a) -> GenT m c) -> GenT m c
```
Comment (effectfully) [inline language-plutus-core/generators/Language/PlutusCore/Generators/Internal/Entity.hs]: i did some generalizations which are just for writing less boilerplate at call sites.

File: language-plutus-core/generators/Language/PlutusCore/Generators/Internal/TypedBuiltinGen.hs [hunk 5]
```diff
@@ -66,7 +67,7 @@ instance (PrettyBy config a, PrettyBy config (Term TyName Name ())) =>
     prettyBy config (TermOf t x) = prettyBy config t <+> "~>" <+> prettyBy config x
 
 attachCoercedTerm
-    :: (MonadQuote m, PrettyDynamic a) => TypedBuiltin Size a -> GenT m a -> GenT m (TermOf a)
+    :: Monad m => PrettyDynamic a => TypedBuiltin Size a -> GenT m a -> GenT m (TermOf a)
```
Comment (effectfully) [inline language-plutus-core/generators/Language/PlutusCore/Generators/Internal/TypedBuiltinGen.hs]: Crossing language boundaries.

File: language-plutus-core/src/Language/PlutusCore/Constant/Apply.hs [hunk 6]
```diff
@@ -104,8 +102,8 @@ instance Enum SizeVar where
     fromEnum (SizeVar sizeIndex) = sizeIndex
 
 -- | Same as 'makeBuiltin', but returns a 'ConstAppResult'.
-makeConstAppResult :: TypedBuiltinValue Size a -> Quote ConstAppResult
-makeConstAppResult = fmap (maybe ConstAppFailure ConstAppSuccess) . makeBuiltin
+makeConstAppResult :: TypedBuiltinValue Size a -> ConstAppResult
+makeConstAppResult = maybe ConstAppFailure ConstAppSuccess . makeBuiltin
```
Comment (effectfully) [inline language-plutus-core/src/Language/PlutusCore/Constant/Apply.hs]: We can ignore `Quote` while evaluating things, because abstract machines not only do not require the global uniqueness condition to hold, but also violate this condition in the process of evaluation (though, we could probably change that, but why).

File: language-plutus-core/src/Language/PlutusCore/Constant/Function.hs [hunk 7]
```diff
@@ -77,31 +77,30 @@ typeSchemeResult (TypeSchemeArrow _ schB) = typeSchemeResult schB
 typeSchemeResult (TypeSchemeAllSize schK) = typeSchemeResult (schK ())
 
 -- | Convert a 'TypedBuiltin' to the corresponding 'Type'.
-typedBuiltinToType :: TypedBuiltin (Type TyName ()) a -> Quote (Type TyName ())
+typedBuiltinToType :: TypedBuiltin (Type TyName ()) a -> Type TyName ()
```
Comment (effectfully) [inline language-plutus-core/src/Language/PlutusCore/Constant/Function.hs]: Crossing language boundaries (compiling types that have a Haskell representation).

File: language-plutus-core/src/Language/PlutusCore/Constant/Make.hs [hunk 8]
```diff
@@ -142,28 +141,27 @@ makeSizedConstant size TypedBuiltinSizedBS   bs  = makeBuiltinBS  size bs
 makeSizedConstant size TypedBuiltinSizedSize ()  = Just $ BuiltinSize () size
 
 -- | Convert a 'Bool' to the corresponding PLC's @boolean@.
-makeBuiltinBool :: Bool -> Quote (Term TyName Name ())
-makeBuiltinBool b = if b then getBuiltinTrue else getBuiltinFalse
+makeBuiltinBool :: Bool -> Term TyName Name ()
```
Comment (effectfully) [inline language-plutus-core/src/Language/PlutusCore/Constant/Make.hs]: Crossing language boundaries.

File: language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs [hunk 9]
```diff
@@ -288,14 +287,15 @@ received evaluator
 
 -- See Note [Semantics of dynamic built-in types].
 -- See Note [Converting PLC values to Haskell values].
+-- Types and terms are supposed to be closed, hence no 'Quote'.
 -- | Haskell types known to exist on the PLC side.
 class KnownDynamicBuiltinType dyn where
     -- | The type representing @dyn@ used on the PLC side.
-    getTypeEncoding :: proxy dyn -> Quote (Type TyName ())
+    toTypeEncoding :: proxy dyn -> Type TyName ()
```
Comment (effectfully) [inline language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs]: Crossing language boundaries.

File: language-plutus-core/src/Language/PlutusCore/FsTree.hs [hunk 10]
```diff
@@ -39,8 +38,8 @@ newtype FolderContents a = FolderContents
 
 -- | A 'PlcEntity' is either a 'Type' or a 'Term'.
 data PlcEntity
-    = PlcType (Quote (Type TyName ()))
-    | PlcTerm (Quote (Term TyName Name ()))
+    = PlcType (Type TyName ())
+    | PlcTerm (Term TyName Name ())
```
Comment (effectfully) [inline language-plutus-core/src/Language/PlutusCore/FsTree.hs]: Those `Quote` were nothing, but an artifact of a ubiquitous `Quote` redundancy.

File: language-plutus-core/src/Language/PlutusCore/TypeCheck.hs [hunk 11]
```diff
@@ -87,7 +87,7 @@ checkKind config ann ty = runTypeCheckM config . checkKindM ann ty
 inferType
     :: (AsTypeError e ann, MonadError e m, MonadQuote m)
     => TypeCheckConfig -> Term TyName Name ann -> m (NormalizedType TyName ())
```
Comment (effectfully) [inline language-plutus-core/src/Language/PlutusCore/TypeCheck.hs]: In principle, we could remove `Quote` from here, but it's much better to do less than to do more. Since we're already monadic here, there is not much benefit (except less calls to `runQuote` at call sites), but if somebody changes something, we can start getting heisenbugs, so we need precise policies on where it's allowed to escape `Quote`.

File: language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Function.hs [hunk 12]
```diff
@@ -264,16 +262,16 @@ getBuiltinFixN n = do
...
 -- >                         , ...
 -- >                         , FunctionDef _ fNn (FunctionType _ an bn) fn
 -- >                         ] =
 -- >     Tuple [(a1 -> b1) ... (an -> bn)] $
 -- >         fixN {a1} {b1} ... {an} {bn}
 -- >             /\(q :: *) -> \(choose : (a1 -> b1) -> ... -> (an -> bn) -> q) ->
 -- >                 \(fN1 : a1 -> b1) ... (fNn : an -> bn) -> choose f1 ... fn
-getBuiltinMutualFixOf :: ann -> [FunctionDef TyName Name ann] -> Quote (Tuple ann)
-getBuiltinMutualFixOf ann funs = do
+getMutualFixOf :: ann -> [FunctionDef TyName Name ann] -> Quote (Tuple ann)
```
Comment (effectfully) [inline language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Function.hs]: It is never allowed to escape `Quote` in a function that receives one or multiple PLC entities as arguments, because those can be open and we can return something open as well.

File: language-plutus-core/stdlib/Language/PlutusCore/StdLib/Meta.hs [hunk 13]
```diff
@@ -3,48 +3,39 @@
...
     :: Type TyName ()
     -> Type TyName ()
     -> Either (Term TyName Name ()) (Term TyName Name ())
-    -> Quote (Term TyName Name ())
-getEitherToBuiltinSum a b (Left  x) =
-    getBuiltinLeft  <&> \left  -> Apply () (mkIterInst () left  [a, b]) x
-getEitherToBuiltinSum a b (Right y) =
-    getBuiltinRight <&> \right -> Apply () (mkIterInst () right [a, b]) y
+    -> Term TyName Name ()
+metaEitherToSum a b (Left  x) = Apply () (mkIterInst () left  [a, b]) x
```
Comment (effectfully) [inline language-plutus-core/stdlib/Language/PlutusCore/StdLib/Meta.hs]: Oops, escaped `Quote` in a function that receives PLC arguments. Bug!

File: language-plutus-core/stdlib/Language/PlutusCore/StdLib/Meta/Data/Tuple.hs [hunk 14]
```diff
@@ -104,8 +103,8 @@ bindTuple ann names (Tuple elTys term) body = liftQuote $ do
 -- | Given an arity @n@, create the n-ary product type.
 --
 -- @\(T_1 :: *) .. (T_n :: *) . all (R :: *) . (T_1 -> .. -> T_n -> R) -> R@
-getBuiltinProdN :: MonadQuote m => Int -> m (Type TyName ())
-getBuiltinProdN arity = do
+prodN :: Int -> Type TyName ()
+prodN arity = runQuote $ do
```
Comment (effectfully) [inline language-plutus-core/stdlib/Language/PlutusCore/StdLib/Meta/Data/Tuple.hs]: We can create a CAF that is an infinite list that contains products of all arities and then simply take the `i`th element of a list, where `i` is an arity. This way we avoid regenerating products and can share their definitions.

File: language-plutus-core/test/Evaluation/even2.plc.golden [hunk 15]
```diff
@@ -1 +1 @@
-(abs a_15 (type) (lam x_16 a_15 (lam y_17 a_15 x_16)))
\ No newline at end of file
+(abs a_0 (type) (lam x_1 a_0 (lam y_2 a_0 x_1)))
```
Comment (effectfully) [inline language-plutus-core/test/Evaluation/even2.plc.golden]: Note that we're doing much fewer renamings now.

File: plutus-core-interpreter/test/DynamicBuiltins/Logging.hs [hunk 16]
```diff
@@ -66,10 +67,6 @@ test_logInts = testCase "logInts" $ do
     let term = runQuote $ do
             let size = 4
                 int4 = TyApp () (TyBuiltin () TyInteger) (TyInt () size)
-            unit         <- getBuiltinUnit
-            unitval      <- getBuiltinUnitval
-            foldList     <- getBuiltinFoldList
-            biEnumFromTo <- getBuiltinEnumFromTo
```
Comment (effectfully) [inline plutus-core-interpreter/test/DynamicBuiltins/Logging.hs]: And don't need to write a lot of junk.

File: plutus-ir/test/TransformSpec.hs [hunk 17]
```diff
@@ -17,12 +17,12 @@ transform = testNested "transform" [
...
-    goldenPir "listFold" (runQuote listFold),
-    goldenPir "monoMap" (runQuote monoMap)
+    goldenPir "listFold" listFold,
+    goldenPir "monoMap" monoMap
     ]
 
-listFold :: Quote (Term TyName Name ())
-listFold = thunkRecursionsTerm =<< do
+listFold :: Term TyName Name ()
+listFold = runQuote $  do
```
Comment (effectfully) [inline plutus-ir/test/TransformSpec.hs]: Oops, stupid bug, noticed it in a test and wanted to emphasize it, but here it is. Fixing.

File: plutus-ir/test/datatypes/listMatchEval.plc.golden [hunk 18]
```diff
@@ -1 +1 @@
-(abs a_42 (type) (lam x_43 a_42 x_43))
\ No newline at end of file
+(abs a_57 (type) (lam x_58 a_57 x_58))
```
Comment (effectfully) [inline plutus-ir/test/datatypes/listMatchEval.plc.golden]: For some reason sometimes we do more renamings.

File: plutus-tx/test/Plugin/recursiveTypes/sameEmptyRoseEval.plc.golden [hunk 19]
```diff
@@ -1,53 +1,53 @@
 (iwrap
-  (lam rec_511 (fun (fun (type) (type)) (type)) (lam spine_512 (fun (type) (type)) [spine_512 [(lam EmptyRose_513 (type) (all out_EmptyRose_514 (type) (fun (fun [List_331 EmptyRose_513] out_EmptyRose_514) out_EmptyRose_514))) [rec_511 (lam dat_515 (type) dat_515)]]]))
-  (lam dat_516 (type) dat_516)
+  (lam rec_353 (fun (fun (type) (type)) (type)) (lam spine_354 (fun (type) (type)) [spine_354 [(lam EmptyRose_355 (type) (all out_EmptyRose_356 (type) (fun (fun [List_173 EmptyRose_355] out_EmptyRose_356) out_EmptyRose_356))) [rec_353 (lam dat_357 (type) dat_357)]]]))
+  (lam dat_358 (type) dat_358)
```
Comment (effectfully) [inline plutus-tx/test/Plugin/recursiveTypes/sameEmptyRoseEval.plc.golden]: But normally fewer.

File: language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Function.hs (hunk 12 above)
Comment (effectfully) [reply language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Function.hs]: It is also never allowed to escape `Quote` in a function that returns type or term variables as a part of its result, because those are just open things.

File: language-plutus-core/examples/Language/PlutusCore/Examples/Data/InterList.hs [hunk 20]
```diff
@@ -39,8 +38,8 @@ We encode the following in this module:
 --
 -- > fix \(interlist :: * -> * -> *) (a :: *) (b :: *) ->
 -- >     all (r :: *). r -> (a -> b -> interlist b a -> r) -> r
-getBuiltinInterList :: Quote (RecursiveType ())
-getBuiltinInterList = do
+interListData :: RecursiveType ()
+interListData = runQuote $ do
```
Comment (effectfully) [inline language-plutus-core/examples/Language/PlutusCore/Examples/Data/InterList.hs]: There isn't much point in escaping `Quote` here, but I also think there isn't much danger, because we return a type (in this situation it's allowed to escape `Quote`) and a function, i.e. a Haskell value (even though it's a function from `Term` to `Term`, which can be potentially dangerous, but then it already was dangerous).

In general, we should avoid reasoning about where it's safe and where it's not safe to escape `Quote`, but I think it's fine to have some corner cases that we thought about seriously.

File: plutus-ir/test/TransformSpec.hs (hunk 17 above)
Comment (effectfully) [reply plutus-ir/test/TransformSpec.hs]: Fixed.

File: plutus-ir/src/Language/PlutusIR/Transform/ThunkRecursions.hs [hunk 21]
```diff
@@ -108,6 +111,9 @@ left with the PLC subset. This would be quite elegant - except that the final "l
 would be quite ugly as it would just have to fail if it saw any PIR features.
 -}
 
+thunkRecursionsTerm :: MonadQuote m => Term TyName Name a -> m (Term TyName Name a)
+thunkRecursionsTerm = PLC.rename >=> thunkRecursionsTermGo
```
Comment (effectfully) [inline plutus-ir/src/Language/PlutusIR/Transform/ThunkRecursions.hs]: @michaelpj, this looks ugly. I'm also not sure if it's needed at all. I can experiment, but I guess it's better if you just tell me.

File: language-plutus-core/src/Language/PlutusCore/TypeCheck.hs [hunk 22]
```diff
@@ -63,7 +63,7 @@ dynamicBuiltinNameMeaningsToTypes
     => ann -> DynamicBuiltinNameMeanings -> m DynamicBuiltinNameTypes
 dynamicBuiltinNameMeaningsToTypes ann (DynamicBuiltinNameMeanings means) = do
     let getType mean = do
-            ty <- liftQuote $ dynamicBuiltinNameMeaningToType mean
+            let ty = dynamicBuiltinNameMeaningToType mean
```
Comment (michaelpj) [inline language-plutus-core/src/Language/PlutusCore/TypeCheck.hs]: Shouldn't there a be a `rename` here? Since we *are* running in `Quote`, and this is a closed-but-not-necessarily-globally-unique thing?

File: language-plutus-core/src/Language/PlutusCore/TypeCheck.hs [hunk 23]
```diff
@@ -87,7 +87,7 @@ checkKind config ann ty = runTypeCheckM config . checkKindM ann ty
 inferType
     :: (AsTypeError e ann, MonadError e m, MonadQuote m)
     => TypeCheckConfig -> Term TyName Name ann -> m (NormalizedType TyName ())
-inferType config = rename >=> runTypeCheckM config . inferTypeM
+inferType config term = rename term >>= runTypeCheckM config . inferTypeM
```
Comment (michaelpj) [inline language-plutus-core/src/Language/PlutusCore/TypeCheck.hs]: imo this is less clear.

File: plutus-ir/src/Language/PlutusIR/Optimizer/DeadCode.hs [hunk 24]
```diff
@@ -28,11 +29,12 @@ removeDeadBindings
     -> Term tyname name a
 removeDeadBindings t =
     let
+        tRen = PLC.runQuote $ PLC.rename t
```
Comment (michaelpj) [inline plutus-ir/src/Language/PlutusIR/Optimizer/DeadCode.hs]: I'd prefer to just document that you need to have run this beforehand.

File: plutus-ir/src/Language/PlutusIR/Compiler.hs [hunk 25]
```diff
@@ -21,7 +22,7 @@ import           Control.Monad
 
 -- | Compile a 'Term' into a PLC Term. Note: the result *does* have globally unique names.
 compileTerm :: Compiling m e a => Term TyName Name a -> m (PLCTerm a)
-compileTerm = (PLC.rename <=< Term.compileTerm <=< thunkRecursionsTerm) . original
+compileTerm = (PLC.rename <=< Term.compileTerm <=< thunkRecursionsTerm) . original <=< PLC.rename
```
Comment (michaelpj) [inline plutus-ir/src/Language/PlutusIR/Compiler.hs]: Again, I would just say that you should run `rename` before you run `compile`.

File: plutus-ir/src/Language/PlutusIR/Transform/ThunkRecursions.hs (hunk 21 above)
Comment (michaelpj) [inline plutus-ir/src/Language/PlutusIR/Transform/ThunkRecursions.hs]: Ditto. I think this *does* need renaming, but it should be done by the caller.

However, at the moment you're effectively running it twice in a row, which is at least wasteful.

Comment (michaelpj) [reply plutus-ir/src/Language/PlutusIR/Transform/ThunkRecursions.hs]: Plus then you don't have to rename the other perfectly well-named function ;)

**Comment by michaelpj** [comment]
Re PIR: I was going to put up a PR that just runs `rename` at the beginning of the pipeline. I think all the transformations need global uniqueness, but should also preserve it, so it's okay to just run it at the beginning. I do have some sympathy for the idea that we should just run it before every pass that needs it  - indeed, given that we do that in the PLC lib now, maybe we should delete the call to `rename` after PIR compilation?

File: language-plutus-core/src/Language/PlutusCore/TypeCheck.hs (hunk 23 above)
Comment (effectfully) [reply language-plutus-core/src/Language/PlutusCore/TypeCheck.hs]: This was an accidental change, I've already fixed it.

File: language-plutus-core/src/Language/PlutusCore/TypeCheck.hs (hunk 22 above)
Comment (effectfully) [reply language-plutus-core/src/Language/PlutusCore/TypeCheck.hs]: I believe no, because I think *kind* checking does not require global uniqueness, because we do not do any type normalization and environments correctly reflect scopes. This needs to be documented, though.

I do plan to add `newtype` wrappers to stress where renaming is necessary and where is not.

File: plutus-ir/src/Language/PlutusIR/Optimizer/DeadCode.hs (hunk 24 above)
Comment (effectfully) [reply plutus-ir/src/Language/PlutusIR/Optimizer/DeadCode.hs]: I can start adding `newtype` wrappers in this PR (comments are completely unreliable), but I'm not sure about this, because then

1. A lot of testing code will need to manually rename things?
2. How often do you have a term that is already globally unique, but such that you haven't called `rename` over it one line above? Preserving global uniqueness is hard, we should only do this when absolutely necessary, it doesn't seem like well-separated PIR transformation passes will suffer from noticeable performance overhead if each of them starts renaming things before processing them. It is way more reliable to just satisfy the invariant rather than to require it to hold. Even when you impose an invariant via a `newtype` wrapper, it's still more reliable to fulfill it automatically, because someone might use the constructor of a data type (like we do with `NormalizeType` for example) or even `coerce` in case the constructor is not exported.

So I believe it's better to rename incoming things rather than outcoming things. This is kinda the essence of the new approach: if you need global uniqueness, then fulfill and maintain it locally, if you need to maintain it. At least I'd start with the local maintenance approach and move to the global approach as an optimization (provided we need this optimization, which I think is questionable).

Comment (michaelpj) [reply plutus-ir/src/Language/PlutusIR/Optimizer/DeadCode.hs]: On reflection I agree. It's probably not too wasteful. In that case I think you can remove the two calls to `rename` bracketing `compile` at least.

Comment (michaelpj) [reply plutus-ir/src/Language/PlutusIR/Optimizer/DeadCode.hs]: However, I still think the newtypes would be helpful!

File: plutus-ir/src/Language/PlutusIR/Compiler.hs [hunk 26]
```diff
@@ -22,7 +22,7 @@ import           Control.Monad
 
 -- | Compile a 'Term' into a PLC Term. Note: the result *does* have globally unique names.
 compileTerm :: Compiling m e a => Term TyName Name a -> m (PLCTerm a)
-compileTerm = (PLC.rename <=< Term.compileTerm <=< thunkRecursionsTerm) . original <=< PLC.rename
+compileTerm = Term.compileTerm <=< thunkRecursionsTerm . original <=< PLC.rename
```
Comment (michaelpj) [inline plutus-ir/src/Language/PlutusIR/Compiler.hs]: You don't need the other one either, since `thunkRecursionsTerm` also does it.

**Comment by effectfully** [comment]
> I still feel anxious about getting the policy right. What about something like a `Closed` wrapper

I tried doing that, but I'm not sure we win much.

> So then you can state a bit more explicitly "why" you're leaving `Quote`: "here I am making a closed term so I'm going to discharge Quote now".

It is *only* safe to call `runQuote` if you have a closed thing. If we ever allow to call `runQuote` over an open term, then you might construct a term that contains variables generated in some another `Quote` (i.e. not current) and you're in trouble. In general, "leaving `Quote` is only allowed if you're constructing a closed thing" sounds like a very sensible and not too intrusive invariant.

> And then in some places you can indicate in the types what's going on.
> 
> ```haskell
> locallyUnique :: Closed (Term TyName Name a) -> Quote (Term TyName Name a)
> locallyUnique = pure . unClosed
> 
> globallyUnique :: Closed (Term TyName Name a) -> Quote (Term TyName Name a)
> globallyUnique = rename . unClosed
> ```

So we have this `Closed` wrapper and just remove it when we don't care about global uniqueness and do something more involved when we do care about global uniqueness. However it means that

1. we write boilerplate in at least 95% of situations (I think it's substantially more than just 95)
2. we still have to be very careful in all other situations

Consider the type checker for example. We have this line there:

```haskell
inferTypeOfBuiltinM (DynBuiltinName ann name) = lookupDynamicBuiltinNameM ann name >>= rename
```

Yes it would be nice to make `lookupDynamicBuiltinNameM` to return a `Closed (Type TyName ())`, so that we are forced to do something about `Closed` and choose between just unwrapping the type or lifting it so that it's globally unique. But

1. it's really easy to repeat what you do all the time (simple unwrapping) in a situation where you need to do something else (lifting)
2. we still have this in the type checker: `lookupVarM name >>= rename`. The type of a variable is not closed, you just have to know that you're in a globally unique context and must rename the type of the variable. You still have to be very careful and manually preserve global uniqueness everywhere in a context where you care about it

There is one advantage of having `Closed`, though. Making something globally unique does not require "marking" if you know that the thing you're lifting is closed:

```haskell
newtype Closed a = Closed
    { unClosed :: a
    } deriving (Show, Eq, Functor, Foldable, Traversable)

instance HasUnique (tyname ann) TypeUnique => Rename (Type tyname ann) where
    -- See Note [Marking].
    rename = through markNonFreshType >=> runDirectRenameM . renameTypeM

instance HasUnique (tyname ann) TypeUnique => Rename (Closed (Type tyname ann)) where
    rename = traverse $ runDirectRenameM . renameTypeM

globallyUnique :: Rename (Closed a) => Closed a -> Quote a
globallyUnique = fmap unClosed . rename
```

But we probably don't care about this optimization too much right now.

Summarizing, while I think that this `Closed` idea is nice and useful, I'm not sure I want to constantly write boilerplate, provided that benefits are not too compelling.

Maybe you have some use cases that I just haven't considered?

**Comment by michaelpj** [comment]
I think maybe we'd just have to try it. I might give it a go.

re: the type variable case, it seems like you're complaining about not having `Open` for things that are known-open. We could do that too!

I think my problem is that `runQuote` doesn't sound like it's related to the open/closedness of the term at all...

**Comment by effectfully** [comment]
> I think maybe we'd just have to try it. I might give it a go.

I'm already doing it, that's how I got the code above and all the objections.

> re: the type variable case, it seems like you're complaining about not having `Open` for things that are known-open. We could do that too!

Then you need not to forget to introduce `Open` somewhere instead of needing not to forget to rename it later. I agree it's a little better though.

What would be even better is to make type variables in the normalizer / types of term variables in the type checker explicitly `Dupable` with the only way to escape `Dupable` being via `rename`. This is much stronger than "here are two ways to escape something, choose wisely" and doesn't seem to have any overhead, so I'm going to do that.

> I think my problem is that `runQuote` doesn't sound like it's related to the open/closedness of the term at all...

Well, it does relate to closedness, because as I'm arguing above you must only call `runQuote` over closed things, otherwise you'll eventually get in an undebuggable trouble.

**Comment by michaelpj** [comment]
> Well, it does relate to closedness, because as I'm arguing above you must only call `runQuote` over closed things, otherwise you'll eventually get in an undebuggable trouble.

No, I mean that when I read it I don't think of closedness so I'm worried I'll forget.

**Comment by effectfully** [comment]
> No, I mean that when I read it I don't think of closedness so I'm worried I'll forget.

Then we can just rename `runQuote` to `runQuoteClosed` :)

**Comment by michaelpj** [comment]
I genuinely think that's a pretty good idea.

**Comment by effectfully** [comment]
No objections on my part.

File: language-plutus-core/examples/Language/PlutusCore/Examples/Data/TreeForest.hs (hunk 2 above)
Comment (effectfully) [reply language-plutus-core/examples/Language/PlutusCore/Examples/Data/TreeForest.hs]: Done.

**Comment by effectfully** [comment]
> What would be even better is to make type variables in the normalizer / types of term variables in the type checker explicitly `Dupable` with the only way to escape `Dupable` being via `rename`. This is much stronger than "here are two ways to escape something, choose wisely" and doesn't seem to have any overhead, so I'm going to do that.

Did that. The result is that I removed all `rename`s from the internals of the normalizer and the type checker, which I think is a really good result. We still have some `rename`s in tests, in the user-facing API of those procedures (which I think is totally fine and doesn't require `newtype` wrappers, because we already separate safe things from unsafe using module boundaries) and in `PlutusCore.hs`, which I deliberately do not touch in this PR as it needs some separate refactoring.

So I do not think `Closed` is important right now. We now have something much better which forces us to rename types/terms where this is required. I.e. instead of separate `Closed` and `Open` we just have a single `Dupable` and use it both the cases and do not provide a `Dupable a -> a` function, only

```haskell
liftDupable :: (MonadQuote m, Rename a) => Dupable a -> m a
liftDupable = liftQuote . rename . unDupable
```

I also feel like the current way `plutus-ir` works is sensible too. I do not see where to put `newtype`s there to make things more reliable, the current approach seems best: you just automatically fulfill the invariant instead of requiring it to hold.

So I propose we merge this PR (once I add a bit more docs) and then improve things separately if we find how.

File: language-plutus-core/src/Language/PlutusCore/Rename.hs [hunk 27]
```diff
@@ -46,3 +48,17 @@ instance (HasUnique (tyname ann) TypeUnique, HasUnique (name ann) TermUnique) =>
 
 instance Rename a => Rename (Normalized a) where
     rename = traverse rename
+
+newtype Dupable a = Dupable
+    { unDupable :: a
+    } deriving (Show, Eq, Functor, Foldable, Traversable)
+
+instance Applicative Dupable where
```
Comment (michaelpj) [inline language-plutus-core/src/Language/PlutusCore/Rename.hs]: `deriving (Applicative, Monad) via Identity`, one day...

Comment (effectfully) [reply language-plutus-core/src/Language/PlutusCore/Rename.hs]: We can actually make it 

```haskell
newtype Dupable = Dupable
    { unDupable :: Identity a
    } deriving (Show, Eq, Functor, Foldable, Traversable, Applicative, Monad)
```

since none of the `Dupable` and `unDupable` are exported anyway.

Comment (effectfully) [reply language-plutus-core/src/Language/PlutusCore/Rename.hs]: Did that. A little bit weird, but concise and sensible, so why not.

**Comment by effectfully** [comment]
Though, we need a `newtype` wrapper whenever we promise to preserve global uniqueness. I'll experiment with this in a separate branch.

Merging.
